src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.cs

```diff
@@ -155,44 +155,61 @@ public IEnumerable<QueryResult> Query(IndexQueryServerSide query, QueryTimingsSc
                             continue;
                         }
 
-                        var result = retriever.Get(document, scoreDoc, _state);
-                        if (scope.TryIncludeInResults(result) == false)
+                        var r = retriever.Get(document, scoreDoc, _state);
+                        if (r.Document != null)
                         {
-                            result?.Dispose();
-
-                            skippedResults.Value++;
-                            continue;
+                            var qr = CreateQueryResult(r.Document);
+                            if(qr.Result == null)
+                                continue;
+                            yield return qr;
+                        }
+                        else if (r.List != null)
+                        {
+                            foreach (Document item in r.List)
+                            {
+                                var qr = CreateQueryResult(item);
+                                if(qr.Result == null)
+                                    continue;
+                                yield return qr;
+                            }
                         }
 
-                        returnedResults++;
-
-                        if (isDistinctCount == false)
+                        QueryResult CreateQueryResult(Document d)
                         {
-                            Dictionary<string, Dictionary<string, string[]>> highlightings = null;
-                            if (query.Metadata.HasHighlightings)
+                            if (scope.TryIncludeInResults(d) == false)
                             {
-                                using (highlightingScope?.Start())
-                                    highlightings = GetHighlighterResults(query, _searcher, scoreDoc, result, document, documentsContext);
+                                d?.Dispose();
+
+                                skippedResults.Value++;
```

Comment:
Paging work before this is called. So that isn't accounted for this.
For multiple results, you are correct, this should be once per document, not result.